frontend/packages/topology/src/layouts/ColaLayout.ts

```diff
@@ -74,133 +108,516 @@ class ColaNode implements webcola.Node {
+
+type ColaLayoutOptions = {
+  linkDistance: number;
+  nodeDistance: number;
+  groupDistance: number;
+  collideDistance: number;
+  simulationSpeed: number;
+  chargeStrength: number;
+  layoutOnDrag: boolean;
+  maxTicks: number;
+  initialUnconstrainedIterations: number;
+  initialUserConstraintIterations: number;
+  initialAllConstraintsIterations: number;
+  gridSnapIterations: number;
 };
 
-export default class ColaLayout implements Layout {
+class ColaLayout implements Layout {
   private graph: Graph;
 
-  constructor(graph: Graph) {
+  private d3Cola: any;
+
+  private forceLink: d3.ForceLink<ColaNode, ColaLink>;
+
+  private simulation: any;
+
+  private options: ColaLayoutOptions;
+
+  private scheduleHandle?: number;
+
+  private scheduleRestart = false;
+
+  private nodes: ColaNode[] = [];
+
+  private edges: ColaLink[] = [];
+
+  private groups: ColaGroup[] = [];
+
+  private nodesMap: { [id: string]: ColaNode } = {};
+
+  private tickCount = 0;
+
+  constructor(graph: Graph, options?: Partial<ColaLayoutOptions>) {
     this.graph = graph;
+    this.options = {
+      ...{
+        linkDistance: 30,
+        nodeDistance: 0,
+        groupDistance: 0,
+        collideDistance: 0,
+        simulationSpeed: 10,
+        chargeStrength: 0,
+        layoutOnDrag: true,
+        maxTicks: 200,
+        initialUnconstrainedIterations: 200,
+        initialUserConstraintIterations: 50,
+        initialAllConstraintsIterations: 150,
+        gridSnapIterations: 50,
+      },
+      ...options,
+    };
+
+    graph
+      .getController()
+      .addEventListener<DragNodeEventListener>(DRAG_NODE_START_EVENT, this.handleDragStart)
+      .addEventListener<DragNodeEventListener>(DRAG_NODE_END_EVENT, this.handleDragEnd);
+
+    this.d3Cola = webcola.d3adaptor(d3);
+    this.d3Cola.handleDisconnected(true);
+    this.d3Cola.avoidOverlaps(true);
+    this.d3Cola.linkDistance(this.getLinkDistance);
+    this.d3Cola.on('tick', () => {
+      if (this.tickCount++ % this.options.simulationSpeed === 0) {
+        action(() => this.nodes.forEach((d) => d.update()))();
+      }
+      if (this.options.maxTicks >= 0 && this.tickCount > this.options.maxTicks) {
+        this.d3Cola.alpha(0);
+      }
+    });
+    this.d3Cola.on('end', () => {
+      action(() => {
+        this.nodes.forEach((d) => d.update());
+        if (this.options.layoutOnDrag) {
+          this.useForceSimulation();
+        }
+      })();
+    });
+
+    this.setupForceSimulation();
   }
 
-  destroy(): void {}
+  destroy(): void {
+    this.graph
+      .getController()
+      .removeEventListener(DRAG_NODE_START_EVENT, this.handleDragStart)
+      .removeEventListener(DRAG_NODE_END_EVENT, this.handleDragEnd);
+
+    this.stopListening();
+
+    this.d3Cola.alpha(0);
+    this.simulation.alpha(0);
+  }
+
+  handleDragStart = (element: Node, event: DragEvent, operation: string) => {
+    // Set the alpha to 0 to halt any ticks that may be occurring
+    this.d3Cola.alpha(0);
```

Comment:
So cola doesn't support `fixed` coordinates?
This is different behavior than the force layout.

Reply:
Cola does have a fixed setting for nodes. However, we are using force simulation for drag as it does a much better job (Cola doesn't really handle it at all). At the start of a drag operation, we set the cola alpha to `0` to stop the cola layout since the user is taking control at that point. Not stopping the cola layout causes the cola and force layouts to compete and nodes to bounce around.

Reply:
Of course! missed the subtlety that this is stopping cola and then regular force is taking over.